File: 383. Ransom Note.cpp
class Solution {
public:
	// unorderd map solution
	bool canConstruct(string ransomNote, string magazine) {
		unordered_map<char, int> letters;   // to store the available letters
		for(int i = 0; i < magazine.size(); i ++)       // record the letters and their corrsponding quantity
			letters[magazine[i]] ++;
		for(int i = 0; i < ransomNote.size(); i ++)      // check the ransom note
			if(-- letters[ransomNote[i]] < 0)
				return false;
		return true;
	}

    // use vector as hashtable: O(m+n), 61.01%, ok
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> letter(26, 0);
        for(auto c: magazine) letter[c - 'a']++;
        for(auto c: ransomNote){
            if(--letter[c - 'a'] < 0)
                return false;
        }
        return true;
    }
};

File: 383. Ransom Note.py
from collections import Counter
class Solution(object):
    
    # hashtable: O(m+n) time, O(1) space
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        cnt = Counter(magazine)
        for letter in ransomNote:
            if letter not in cnt or cnt[letter] <= 0:
                return False
            cnt[letter] -= 1
        return True
        